### ISSUE TITLE
Default config (state) dir on Unixes should be ~/.local/state/syncthing ($XDG_STATE_HOME/syncthing)

### ISSUE THREAD

**Issue by hholst80** [syncthing/syncthing#9178]
I would like to be able to sync my .config directory with syncthing without it breaking across multiple machines.

The reason it breaks now is due to the fact that the local installation stores database and certificates and such local deployment data under the same directory.

I think sharing .config wholesale is possible without having to exclude or move .config/syncthing if we consider what is sharable data and what is local data.

I will add my findings here but please jump in if you either have explored this already or you are interested in exploring this.

https://wiki.archlinux.org/title/XDG_Base_Directory

**Comment by tomasz1986** [comment]
Duplicate of https://github.com/syncthing/syncthing/issues/8102?

**Comment by hholst80** [comment]
It is related to the issue, but no attempt was made to solve it.  I think that if we involve the same people again, we will reach the same conclusions.

If an application uses XDG conventions then it must adhere to the conventions.  What is stored in .config/syncthing right now is a mix of things that XDG defines.  Syncthing acknowledges this (config *and* data = home directory), but out of the box it opts to a imperfect solution that will lead to *recoverable data loss* in certain scenarios.

ST will warn for this, after the fact, but then it is typically too late.

A sensible solution is to use --home=.local/share/syncthing which so far seems to solve most things. There are parts of the config that are sharable between machines, but there are others that are not.  An ambitious change would be to split up the config into parts that are local and those that can roam across multiple machines.

Migrating this won't be easy but until this is resolved ST will continue to break in the use case where users share .config across multiple machines. Which I suppose is a common use case.

Migration path:

Start using `.local/share/syncthing` as primary location; if not found revert to `.config/syncthing` and warn user about deprecation and link to a post which describes how to move `.config/syncthing` to `.local/share/syncthing`.

**Comment by calmh** [comment]
We're not going to change where things are stored by default, but you have --data and --config to accomplish what you want.

**Comment by hholst80** [comment]
Why should users have to opt in not to have their data destroyed? Can you give a rationale so users finding this issue later will understand why it makes sense?

**Comment by calmh** [comment]
Syncthing does what you want if you take minimal steps to indicate that the system wants to use XDG, for example by setting `XDG_DATA_HOME`

https://docs.syncthing.net/users/config.html

> The database is stored either in the same directory as the config (usually the default), but may also be located in one of the following directories (Unix-like platforms only):
>
> - If a database exists in the old default location, that location is still used.
> - If $XDG_DATA_HOME is set, use $XDG_DATA_HOME/syncthing.
> - If ~/.local/share/syncthing exists, use that location.

**Comment by acolomb** [comment]
I agree, however, that the Right Thing to do for Syncthing would be to handle `$XDG_CONFIG_HOME` as indicated in the XDG spec.  People expect to be able to freely copy their "configuration" (i.e. `~/.config/*`) to another machine without causing harm.  Witnessed here for example with catastrophic consequences: https://forum.syncthing.net/t/copying-old-syncthing-config-led-to-data-loss/20976

There are clear rules laid out in the spec regarding which default values to fall back to if the variables are not defined.  Thus our requirement of needing to define the variable to make it "adhere to the spec" is actually not doing that.  Instead, we use a path by default that provokes certain expectations (sharability), but put stuff there that certainly doesn't meet them.

IMO the right location for the database would be under `$XDG_STATE_HOME/syncthing`, respectively `~/.local/state/syncthing/`.  The device ID (certificate and key) is kind of "configuration", but definitely not sharable, thus should probably live there as well.  Even the XML configuration is not really sharable, because it's not holding just user preferences, but sharing relationships that are specific to this user's instance running on exactly this device.

They're also not "data" as the XDG spec generalizes, since there is no evidence that `.local/share/` should *not* be shared among devices for the same user.  That's where the XDG model doesn't really suit Syncthing's needs, but I think we made the wrong choice using `~/.config` by default.  By the way, the `--config` and `--data` separation is orthogonal to the issue, because it is intended to flexibly place the DB on a storage medium best suited for the I/O access patterns.

So I'd favor changing the default for new installations to keep both config and database in the `$XDG_STATE_HOME` paths, avoiding the reported sharing problems.  Keeping the old paths' fallback logic intact of course.  It would also better cover the common use-case of wanting to share the `~/.config` hierarchy completely, without needing ignore patterns.

**Comment by calmh** [comment]
As far as I understand the XDG "standard" was never intended to apply to things like user apps (it's for reference by other XDG stuff to decide where to place menu items and whatnot) and doesn't say anything about user-*and-machine*-specific state like Syncthing keeps. Then people got annoyed by dotfile clutter and latched on to whatever they could find that claimed to be a standard and here we are. Trying to shoehorn our stuff into some directory plan that was never designed for it was a mistake to begin with.

**Comment by acolomb** [comment]
It was really a double-mistake.  If the XDG path is unfit for the purpose, we should have chosen (or kept?) `~/.syncthing/`.  But when jumping on the XDG train, we chose a subdirectory there which does not align well with how Syncthing uses it, and `$XDG_STATE_HOME` would have been more appropriate.  I definitely agree that the XDG spec does not apply well to Syncthing as a user-and-machine specific service, but I wouldn't call it ill-conceived overall.

Since we are here and using a path (by default) for which certain rules are laid out and widely accepted, further troubles are to be expected if we don't change that.

**Comment by calmh** [comment]
It's a series of mistakes. We started with ~/.syncthing indeed, everything since that point was driven by someone yelling about xdg but apparently no one understanding it, which I can't say I blame anyone for since it isn't written for us. The "state" stuff being discussed now seems to be some new addition and again I don't really see that it applies very well to us.

**Comment by acolomb** [comment]
Not very well, but at least it's the best one of the defined options.  Whatever we do, staying in `~/.config` is harmful, as witnessed in the forum topic.  So either back out from XDG completely, or switch to `$XDG_STATE_HOME`, to avoid breaking people's expectations from the paths used.

I do see some value in not cluttering the dotfile space further, so the second option seems more appealing to me.  Would need to see consensus on that before starting to work on it, though.  And right now I can't promise to find time soon for it.

**Comment by calmh** [comment]
So, trying to read the tea leaves:

> $XDG_STATE_HOME defines the base directory relative to which user-specific state files should be stored. If $XDG_STATE_HOME is either not set or empty, a default equal to $HOME/.local/state should be used.
> The $XDG_STATE_HOME contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in $XDG_DATA_HOME. It may contain:
> - actions history (logs, history, recently used files, …)
> - current state of the application that can be reused on a restart (view, layout, open files, undo history, …)

Our stuff is certainly "not portable enough", if that means portable to a different machine. I would not argue it's unimportant though, if one wants Syncthing to work.

The current "data" definition seems to be:

> $XDG_DATA_HOME defines the base directory relative to which user-specific data files should be stored. If $XDG_DATA_HOME is either not set or empty, a default equal to $HOME/.local/share should be used.

This is from the "environment variables" section of https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html, are you guys reading something else that has actual definitions of what's supposed to be in these directories, where are the specific expectations on these directories coming from?

**Comment by acolomb** [comment]
There is the philosophical analogy to `/usr/share/` mentioned in https://wiki.archlinux.org/title/XDG_Base_Directory#User_directories, which is of course not part of the standard though.  Following that analogy leads to the Filesystem Hierarchy Standard at https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s11.html#purpose26, which describes the platform- and machine dependency in much more detail.

**Comment by acolomb** [comment]
> Of course, we "expose" stuff that would be under /var/lib-equivalent, by documenting it, which I guess is then wrong. Maybe we should remove our config docs and consider it private...

No, please not.  We'd need a separate developer documentation then for people interacting with it outside of the GUI itself.  Such as wrappers and people making scripted deployments for example.  That would be worthwhile in itself (together with the API and protocol specs), but currently not within scope or the available developer capacity :-)

**Comment by calmh** [comment]
I was being facetious. We should certainly not remove our docs.

**Comment by calmh** [comment]
So, for the record, I consider the XDG standard badly written, ill-defined, misunderstood, and misapplied. However, people keep insisting that it should apply, and that we should follow it, despite it also being a moving target. Fine, we can do that. On the plus side, it cleans things up in the sense that it consolidates all our stuff in one directory again, and also clarifies that users should not care about or fiddle with the files in there. That's good. Carry on.

**Comment by piotr-dobrogost** [comment]
> (…) despite it also being a moving target

Just wanted to note that most useful specifications are live meaning they adapt to ever changing environment they live in and they are being refined to better convey the ideas behind them. Calling it a _moving target_ is not fair.

**Comment by piotr-dobrogost** [comment]
You could consider unlocking https://github.com/syncthing/syncthing/issues/8102 and relocking it again which I believe should result in that issue getting a comment with a link to this one (which hasn't been created in the first place because of the lock).

**Comment by acolomb** [comment]
Right, already forgot that we had discussed this back there.  You gave the link right here, think that's enough.  #8102 was already closed and we did reference two other issues in the PR.

**Comment by ernstki** [comment]
Hi all, and thanks for your efforts on this.

As a new Syncthing user, not really knowing any better and for reasons beyond the scope here, I _did_ want to locate and tinker with the config.

It _was_ a bit confusing not finding config files in `~/.syncthing` or `~/.config/syncthing`, and while I'm not entirely ignorant of the XDG "standard" directories, `~/.local/state` was completely new to me. But the reasoning here makes sense, even if you have to have read between the lines of the spec to have arrived at `state` vs `local` or `config`.

I can definitely commiserate with the project maintainers regarding the pressure from vocal user(s) to follow this or that convention, however ill-advised. All the same, I appreciate the care with which you've discussed the issue here and made the decision you did, even if it's imperfect, in order to prevent data loss in some rare cases.

Cheers.

**Comment by dragomang87** [comment]
I just want to add another example of possible mess due to this choice of folders (also new user here): I recently got fed up with various problems with various applications that were due to config and cache rot, so I decided to delete everything in .local/state and .cache and what not that I thought should not impact applications other than rebuilding needed files.

Needless to say, it fixed the problems but I was surprised to open syncthing and see it completely reset (now I need to set up this machine again from scratch and decide whether to try and "reproduce" the machine so the other machines think it came back online or just set it up as new again, which means removing the old deleted one from the other machines).

Just because XDG or anyone defines a standard it doesn't mean they get first choice of word selection for their folders.
If they meant `.config` to mean "files related to menus and common stuff used by windows managers", then they initiated the mess even if it was involuntary. "config" means "config(uration)", that's it, they should have called it ".desktop" maybe.
Similarly, "state" means "state", not config, if I erase "state" I expect not to erase "config".

The messy "abuse" of the XDG specs by "people who don't understand it" is sometimes also the use those folders should have had to begin with.

At least with `~/.syncthing" it was signaling something like "syncthing does not fit any of the common practices or specifications and required special treatment".

**Comment by acolomb** [comment]
> I decided to delete everything in .local/state and .cache and what not that I thought should not impact applications

Well, that thinking was wrong, unfortunately.  "Application state" can mean a lot of things, but if you remove it, then it should be no surprise to find the app back at the default (empty) state.  Of course it can be argued that the device identity and folder paths and connections to other devices is "configuration" data.  But in this case, the common path `~/.config/` caused people to expect *shareable* configuration data, which is pretty harmful in Syncthing's context.  It's always a trade-off between different people's expectations based on some standard that is often the only documentation of expected conventions, even if its assumptions are arguable.

**Comment by dragomang87** [comment]
Indeed, totally my mistake to own, I am trying to point out that there is a "common" expectation that has built on people regarding the use of words like config, state, cache, etc which you cannot escape and that even syncthing has fallen trap of.

Syncthing has a specified set of options that has been explicitly decided is to be stored in `config.xml` and described in the documentation as config, which we are now being told does not belong in `config`. It was not an issue when it was in `~/.syncthing`, it becomes an issue when it is decided to put "config" not in "config".

If the various files were called "file-index.db", "server.-databasexml", "device-database.xml", etc then you don't have the same issue, it would be a way of saying "sorry, Syncthing does not have configuration, it only has working state that is not easily copied", but as long as it is called "config" you will always get the question "why is config not in config?", which will never be a real question, it will always be a polite statement of "honestly, it doesn't really matter who specified what, once you decide what config is, config should be in config, feel free to decide if the mistake was calling it config or not putting it in config".
